[All] Refactor DataTypeInfo system.

damienmarchal opened the pull request:
This PR is a step-by-step version of the experiment done in https://github.com/sofa-framework/sofa/pull/1552

Currently this PR shouldn't breaking too much, the CI will tells us.

Two test are failing...but this is on purpose to keep a track the breaking parts. 
The failure results from the unification of the naming schemes. 
Fixing the tests is easy, just replace the old string with the new one (int->i).
Nevertheless I suggest to fix that when whole refactoring will be over.

Compared to #1552 
- In this PR I'm re-organizing the existing Sofa code without touching its logic. The files in typeinfo/models/* are cut & past from DataTypeInfo.h. The files in typeinfo/TypeInfo_XXX.h are the DataTypeInfo<> specialization that also come from DataTypeInfo.h
- Some breaking changed are introduced to unify the naming scheme. The 'short' naming scheme is following the standard for python.array and was very close to what we already have in Sofa. 

- In this PR I don't use at all the TypeInfoRegistry (expect in the test part), this is on purpose so a future PR will be able to emphasize the change on the Sofa code base (in Data.h). 
- In this PR I don't register typeinfo the registry (because it is not used yet), but I added the 'empty' files for that...it is in typeinfo/TypeInfo_xxxxx.cpp 
 
What will be done in a future PR:
- 1. fill the registry with types we are using and replace the use of VirtualTypeInfo<T>::get() in Data.h by TypeInfoRegistry::Get(T).
- 2. remove step by step the static type info and there related #includes files. 
- 3. rename all the method/enum in the static & dynamic system to unify with official naming convention. 
     a. static function must start with an UpperCase
     b. function returning boolean must start with a verb like IsValidTypeInfo() instead of ValidTypeInfo() 
     c. add a getName()/GetName() to replace name().
- 4. replace the enum involved in the static type info system with constexpr. 

EDIT: To satisfy popular demand here is more or less the content of the PR.  
[Sofa.pdf](https://github.com/sofa-framework/sofa/files/5553887/Sofa.pdf)
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
[SofaDefaultType] move AbstractTypeInfo from DataTypeInfo.h to AbstractTypeInfo.h

damienmarchal committed:
[SofaDefaultType] add setCompilationTarget (resp. get) in AbstractTypeInfo.h

This allows us to track where the typeinfo has been generated.

damienmarchal committed:
[SofaDefaultType] add TypeInfoID, TypeInfoRegistry, and TypeInfoRegistryTools

TypeInfoRegistry to store each defined TypeInfo and retrive it in constant time
using a TypeInfoId;

damienmarchal committed:
{SofaDefaultType] move DataTypeInfo and VirtualTypeInfo in their own .h

DataTypeInfo moves to DataTypeInfo.h
while VirtualTypeInfo is renamed to DataTypeInfoDynamicWrapper. The renaming was done to to make more clear
its role is to generate a dynamic version of a DataTypeInfo.

fredroy committed:
Fix compilation for MSVC

damienmarchal committed:
Merge pull request #1 from fredroy/pr-refactor-datatypeinfo-step1

Fix compilation for MSVC

damienmarchal committed:
[SofaKernel] Split DataTypeInfo.h into independant TypeInfo_*

damienmarchal committed:
[FixUp] Add missing include

fredroy commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/Vec.h:
```diff
@@ -476,7 +476,7 @@ class Vec : public helper::fixed_array<real,N>
     Vec<N,real> mulscalar(real2 f) const
     {
-        static_assert(DataTypeInfo<real2>::ValidInfo && DataTypeInfo<real2>::Size==1, "");
+        //static_assert(DataTypeInfo<real2>::ValidInfo && DataTypeInfo<real2>::Size==1, "");
```
static_assert(std::is_arithmetic< real2 >::value); ?

damienmarchal replied:
Thank for the tip,
 as this is outside of the scope of the PR I let that for future work for a volunteer.

jnbrunet replied:
Hey @damienmarchal ,

This PR is removing those static assertions so it does in my opinion fall into its scope... 

Either remove completely those lines by indicating in the PR description why they are no longer needed, or replace them by an alternative such as the one suggested by @fredroy .

damienmarchal replied:
EDIT: So finally I refactored Vec.h to remove all the static_assert... (and I also renamed the templated name because is is misleading...as we can have Vec of integers)...and no my god...please stop ...a cannot prevent myself from refactoring Vec.....arg.....

jnbrunet replied:
Thanks a lot @damienmarchal , it makes much more sense now.

damienmarchal committed:
[SofaKernel] Add full test in TypeInfoRepository_test.cpp

damienmarchal committed:
[SofaKernel] fix compilation issues because if missing includes.

damienmarchal committed:
[SofaDefaultType] Add NoTypeInfo to handle cases when there is not type info at all

damienmarchal committed:
[SofaPython] fix conversion from std::string to char*

damienmarchal committed:
[SofaDefaultType] Replace char* by std::string in TypeInfo_*

damienmarchal committed:
[all] Replace the use of not correct use of DataTypeName with the one DataTypeInfo.

damienmarchal committed:
[SofaDefaultType] Add new test TypeInfoRepository to detect cross translation unit behavior.

damienmarchal commented:
Now it compiles let's try the full damages: [ci-build][with-scene-tests][with-regression-tests]

damienmarchal committed:
[SofaPython] add new naming convention for type to SofaNumpy.py

damienmarchal committed:
[SofaDefaultType] remove un-used file typeinfo/TypeInfo.h

damienmarchal commented:
@fredroy asked for a second review for #1605. 
Can someone do it because want to have the first step done and the content of the  PR is now more 20 day old.

epernod commented on applications/plugins/SofaPython/python/SofaPython/SofaNumpy.py:
```diff
@@ -19,6 +19,16 @@
     'unsigned int': ctypes.c_uint,
     'long': ctypes.c_long,
     'unsigned long': ctypes.c_ulong,
+    'd': ctypes.c_double,
```
this is needed or some changes from other work?

damienmarchal replied:
No this is part of the PR. In the PR some of the typeinfo names are replaced with their short name so it we need to update  the use cases in the code base to that that changes  into account.

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/typeinfo/models/IncompleteTypeInfo.h:
```diff
@@ -0,0 +1,95 @@
+    template<typename T>
+    static void setValue(DataType& /*data*/, sofa::Size /*index*/, const T& /*value*/){}
+
+    static void getValueString(const DataType& /*data*/, sofa::Size /*index*/, std::string& /*value*/){}
```
SOFA_UNUSED(data) :)

epernod replied:
a lot of commented inputs in the xxTypeInfo. I let you see if you want to clean that in this PR or not.

damienmarchal replied:
Thanks for the comments, well given that was already there in the existing code...and I have no desire to spend time on cosmetic cleaning I will not touch that in the PR.

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/typeinfo/models/FixedArrayTypeInfo.h:
```diff
@@ -0,0 +1,205 @@
+    {
+        if (!FixedSize)
+        {
+            size /= DataType::size();
```
I know this is not related to this PR, but this block seems weird.

damienmarchal replied:
Once the PR is merged feel free to clean the sofa code :)

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/typeinfo/TypeInfo_Integer.h:
```diff
@@ -0,0 +1,101 @@
+{
+    static const std::string GetTypeName() { return "unsigned long long"; }
+    static const std::string name() { return "Q"; }
+};
```
there are some diff with what you use in python SofaNumpy.py, not a problem?

damienmarchal replied:
Yes, I didn't wanted to add new feature to SofaNumpy...but I will to avoid mismatch.

damienmarchal replied:
I have added the extra type in SofaPython

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/Vec.h:
```diff
@@ -23,8 +23,8 @@
 #define SOFA_DEFAULTTYPE_VEC_H
 
 #include <sofa/helper/fixed_array.h>
+#include <sofa/helper/logging/Messaging.h>
```
I don't see logs in the file.

damienmarchal replied:
Probably in other files.

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/TypeInfoRegistry.h:
```diff
@@ -0,0 +1,63 @@
+    NONE,
+    PARTIAL,
+    COMPLETE,
+    ALL
```
what is the diff between Complete and all ?

epernod replied:
ok saw it I think. ALL == (COMPLETE && VALID) ?

damienmarchal replied:
yep

jnbrunet replied:
Could you add some documentation here? What does these typeinfo type mean? What a valid and non valid typeinfo?

damienmarchal replied:
Sure

epernod commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/TypeInfoRegistry.cpp:
```diff
@@ -0,0 +1,109 @@
+
+    if( typeinfos[id] != NoTypeInfo::get() )
+    {
+        if( typeinfos[id] != info && info->ValidInfo())
```
not clear to me here how you could know if a new info is more valid than registered info (also flagged as valid)
So I would suggest a msg_warning rather than msg_info

damienmarchal replied:
When we set a valid type info this override the other one only and only if the new one is valid (so we never overrides a typeinfo with a new one having smaller level of validity()). I consider this the expected behavior so I'm not sure I want to issue warning. I would probably use a warning if the other way around was possible: replacing a valid type info with an invalid one...but this case does not happens.

jnbrunet commented on SofaKernel/modules/SofaDefaultType/SofaDefaultType_test/TypeInfoRepository_test.cpp:
```diff
@@ -0,0 +1,80 @@
+
+TEST(TypeInfoRegistry, dump)
+{
+    TypeInfoRegistryTools::dumpRegistryContentToStream(std::cout, TypeInfoType::NONE);
```
What's the test here?

Also, I think it would be a good idea to avoid the use of cout and cerr for tests (unless we are debugging) since it clogs the test results.

damienmarchal replied:
Well the test here consist in executing the code (with a good percentage of code coverage) without crashing (which is actually far better than a lot of existing sofa code ;)) 

EDIT: more seriously I will replace std::cout with std::stringstream and check it is not empty.

damienmarchal replied:
Done.

jnbrunet commented on SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/AbstractTypeInfo.h:
```diff
@@ -0,0 +1,181 @@
+
+namespace sofa::defaulttype
+{
+/** Information about the type of a value stored in a Data.
```
Very good documentation :+1:

damienmarchal committed:
[plugins/SofaPython] Add new array type to support every registered typeinfo.

damienmarchal committed:
[SofaDefaultType] Use a single TypeInfoID to avoid mismatch between the registery and the Id.

damienmarchal committed:
[SofaKernel] Implement BaseType() and ValueType() without the need of overloading.

damienmarchal committed:
[SofaDefaultType] Removale of the DefaultTypeInfo as fallback for DataTypeInfo

damienmarchal committed:
[SofaDefaultType] Add multiple tests.

damienmarchal committed:
[SofaDefaultType] Add a new class of TypeInfo called NameOnlyTypeInfo

damienmarchal committed:
[SofaDefault] Improve the TypeInfo auto registration mechanism

Now any used type is registered with type NameOnlyType
instead of NoTypeInfo so we can make name queries and equivalence
testing between them.

damienmarchal committed:
[SofaKernel] Update Type Info related tests.

damienmarchal committed:
[SofaDefaultType] Remove broken behavior to make the PR be non breaking.

damienmarchal committed:
[SofaDefaultType] Add SOFA_DEFAULTTYPE_API in class declaration

damienmarchal committed:
Revert "[all] Replace the use of not correct use of DataTypeName with the one DataTypeInfo."

This reverts commit 110026391f1b6965273339f554e2fd04ab5d762c.

damienmarchal committed:
[SofaDefaultType] Fix typo with Sofa_DEFAULTTYPE_API

damienmarchal committed:
[SofaDefaultType] Clean Vec.h to remove the static_assert.

damienmarchal committed:
[SofaDefaultType] Add missing SOFA_DEFAULTTYPE_API in wrapper.h

damienmarchal committed:
[image] All ImageTypeInfo are now inheriting from BaseImageTypeInfo

So we can test & detect parenting relationship through their typeinfo.

damienmarchal commented:
To fix the last test I had to be able to detect that some data are child of BaseImage. So I have added inheritance mecanism in the DataTypeInfoWrapper.

damienmarchal committed:
Update DataTypeInfoDynamicWrapper.h

fredroy committed:
Fix windows compilation

damienmarchal committed:
Merge pull request #2 from fredroy/pr-refactor-datatypeinfo-step1

Fix windows compilation part 2

damienmarchal committed:
Update AbstractTypeInfo.h to add SOFA_DEFAULTTYPE_API
